README
                             pwauth 2.3.12~mwp

This is a fork of Jan Wolter's pwauth.  Jan Wolter passed away on
January 1, 2015.  Development was moved to https://github.com/phokz/pwauth.
That repository has not been updated since August 2018, and release 2.3.11
was still unreleased as of January 2021.

This fork incorporates all of the outstanding pull requests on
https://github.com/phokz/pwauth.  Support for autoconf and building debian
packages building has also been added.  Pull requests and contributors
are welcome.


                             Author: Jan Wolter

                       http://www.unixpapa.com/pwauth/

Pwauth is a conceptually a simple program.  You run it, giving it a login
and a password, and it returns a status code telling whether or not that
login/password is valid.  It is designed to be combined with mod_auth_external
(or mod_authnz_external) and Apache to give reasonably secure HTTP
authentication from a Unix password file, though it can be used in other ways
too.

Mod_auth_external and mod_authnz_external are available from
http://www.unixpapa.com/mod_auth_external/

Pwauth ends up being slightly more complex because of the lack of consistancy
in the way different versions of Unix do authentication.  It includes code
for doing low-level authentication in most different versions of Unix.  It
also can be configured to use one higher-level interface to authentication,
PAM.  All configuration is compiled in, because in typical applications
this program runs very frequently (on every web hit on a protected page),
so the cumulative overhead of reading a config file on every run would be
substantial.

I believe that mod_auth_external, with the included pwauth program, is the
most secure method for doing web authentication out of unix shadow password
systems. Mod_auth_pam or mod_auth_system can also do this, but since they
are internal authenticators, they will only work if you make the shadow
password file readable to the http server. This means that if there are
any exploitable vulnerabilities in the http server, then it may be possible
for people to grab a copy of your shadow password file. Worse, any CGI
program on your system which is not run under suExec or cgiwrap also has
read access to your shadow password database, and any bugs in these might
also expose your entire password database. When mod_auth_external and pwauth
are used, neither the http server nor any CGI programs are given access to
the shadow database. Only the "pwauth" program does. Since it is a small
and simple program, it is much easier to assure that it does not have
security weaknesses.

Having said that, authenticating from a Unix password file is an idea that
many sensible people find seriously questionable.  See Apache's FAQ
(http://httpd.apache.org/docs/misc/FAQ-G.html#passwdauth) for a overview
of some of the issues.  Pwauth has features that can address most of the
arguments made here, if correctly configured, but you need to be aware of
the issues and extremely careful.  I've used it for many years without
problems on systems that are under almost continuous assault by hackers,
but none of those systems are at all typical in their security requirements.
You should think hard about using this software and proceed with caution.

Installation instructions are in the INSTALL file.  The FORM-AUTH file
discusses using this in form-based authentication applications.  Configuration
information is in the comments in the "pwauth.h" file.

Versions of pwauth before version 2.2.8 were distributed as part of the
mod_auth_external distribution.

Author:  Jan Wolter  http://www.unixpapa.com/


pwauth/CHANGES
Pwauth Change Log
=================

VERSION 2.3.12~mwp - Jan 12, 2021
  - Import and tweak Debian build directory.
  - Support GNU autoconf.
  - Set PAM_RHOST item from environment variables.
  - More cleanup regarding headers, declarations, etc.
  - Fork from https://github.com/phokz/pwauth.

VERSION 2.3.11 -
  - Don't define "uid" variable if SERVER_UID is not defined.
  - Removed old-style strchr prototype.
  - Clarified some instructions in config.h.

VERSION 2.3.10 - Oct 5, 2011
  - Changed the serialized sleep code in snooze.c to use fcntl() locking
    instead of flock() locking.  Fcntl() locking is a POSIX standard and
    is likely to work better on more systems, notably including Solaris
    which doesn't seem to support flock() at all any more.
  - Minor fixes to typos in various documentation.

VERSION 2.3.9 - May 2, 2011
  - Add AUTHENTICATE_AIX option for authenticating via AIX's authentication
    configuration system.  Thanks to Hans Dieter Petersen of the University
    of Bonn for this implementation.
  - Renamed PAM_OS_X configuration setting to PAM_OLD_OS_X since it only
    is needed for OS X 10.5 and older.
  - Rearranged ifdef's so that undefining SLEEP_LOCK actually completely
    disables the sleep-on-failure behavior.
  - Minor documentation fixes

VERSION 2.3.8 - Apr 3, 2009
  - Undefining SERVER_UIDS now disables the runtime uid check.  Documentation
    added to suggest using this, together with group execution permissions on
    the binary, to create a group for users who can run pwauth.  Thanks to
    Adi Kriegisch <adi@kriegisch.at> for suggesting this.
  - Return a distinct status code if authentication fails because we are not
    running as root.  This is currently only done for SHADOW_SUN, SHADOW_BSD,
    SHADOW_AIX, and SHADOW_MDW.  It's just to help confused installers
    figure out why things aren't working.
  - Warn installers that they may need to install PAM development packages.

VERSION 2.3.7 - Jan 9, 2009
  - DOCUMENTATION FIX ONLY
  - Corrected erroneous AuthBasicProvider command in INSTALL file.

VERSION 2.3.6 - May 19, 2008
  - Add PAM_OS_X option.
  - Clarified comments in config.h.
  - Replace wildly obsolete inclusion of strings.h with inclusion of string.h

VERSION 2.3.5 - Dec 17, 2007
  - Fixed return codes from AIX and HPUX versions (thanks to Paul Marvin for
    finding this bug).

VERSION 2.3.4 - Nov 11, 2007
  - Fixed PAM_SOLARIS define.

VERSION 2.3.3 - Sep 1, 2007
  - Don't allow logins during inactive period after password expiration.

VERSION 2.3.2 - Feb 19, 2006
  - Update documentation to discuss usage with mod_authnz_external.
  - Update documentation to discuss use of mod_authz_unixgroup instead of the
    unixgroup script.
  - Drop "development release" notation.

VERSION 2.3.1 - Jan 10, 2005
  - Fix the checks for expired passwords and expired accounts for
    LOGIN_CONF_OPENBSD configurations.
  - Fix the handling of the pam_message argument to the conversation function
    for Solaris.  The old handling was right for Linux PAM and OpenPAM, but
    not for Solaris.  However, the bug occurs only when the PAM modules passes
    more than one prompt to the conversation function, which should probably
    never happen.

VERSION 2.3.0 - Sep 28, 2004
  - Status code values returned by pwauth have changed.  0 is still success,
    of course, but there is a much wider range of non-zero error codes returned
    to indicate different causes of login failure.
  - Pwauth now checks for /etc/nologin file by default.  Undefine
    NOLOGIN_FILE in config.h if you don't want this behavior.
  - Pwauth now checks if an account is expired and refuses logins if it is.
    Undefine CHECK_LOGIN_EXPIRATION in config.h if you don't want this
    behavior.
  - Pwauth now checks if an account's password has expired and refuses logins
    if it is and if logins are supposed to be disabled when the password has
    expired.  Undefine CHECK_PASSWORD_EXPIRATION if you don't want this
    behavior.
  - Added support for authenticating through login.conf interface on OpenBSD.
    Support for login.conf systems on other versions of Unix is not yet here.
  - Added support for OpenBSD failure logs.
  - Source code split into multiple files.
  - Added 'checkfaillog' program which CGIs can run to report/reset failures
    and admins can run to reset failure counts.

VERSION 2.2.8 - Sep 25, 2004
  - First separate distribution of pwauth.  This version is identical to
    the version in the mod_auth_external version 2.2.8 package, except for
    repackaging and slight modifications to the documentation.

Versions of pwauth previous to version 2.2.8 were distributed as part of
the mod_auth_external package, and change-log information is included in
pwauth change log.


auth_pam.c
/* =======================================================================
 * Copyright 1996, Jan D. Wolter and Steven R. Weiss, All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The names of the authors must not be used to endorse or
 *    promote products derived from this software without prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
 * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * =======================================================================
 */

#include "pwauth.h"

#if defined(PAM_SOLARIS_26) || defined(PAM_SOLARIS) || defined(PAM_OS_X)
#define PAM
#endif /* PAM_SOLARIS_26 or PAM_SOLARIS */

#ifdef PAM
#ifdef NEED_UID
#include <pwd.h>
#endif
#ifdef PAM_OLD_OS_X
#include <pam/pam_appl.h>
#else
#include <security/pam_appl.h>
#endif
#endif

#ifdef PAM
/* ========================= PAM Authentication ======================== */
/*         Based on version by Karyl Stein (xenon313@arbornet.org)       */
/*            and on parts of mod_auth_pam.c by Ingo Lutkebohnle         */

/* Application data structure passed to PAM_conv: */

struct ad_user {
	const char *login;
	const char *passwd;
	};

/* The pam_unix.so library in Solaris 2.6 fails to pass along appdata_ptr
 * when it calls the conversation function.  So we use a global variable.  */

#ifdef PAM_SOLARIS_26
struct ad_user user_info;
#define PAM_SOLARIS
#endif /* PAM_SOLARIS_26 */

/* PAM_CONF - PAM Conversation Function.  Called by PAM to get login/password.
 * Note that "appdata_ptr" is really a "struct ad_user *" structure.
 */

#ifdef PAM_SOLARIS
/* In Solaris PAM, msg is a pointer to a pointer to a size num_msg array of
 * pam_message structures.
 */
# define msgi(i) ((*msg)[i])
#else
/* In Linux PAM and OpenPAM, msg is a pointer to a size num_msg array of
 * pointers to pam_message structures. 
 */
# define msgi(i) (*(msg[i]))
#endif

#ifdef PAM_SOLARIS_26
int PAM_conv (int num_msg, struct pam_message **msg,
     struct pam_response **resp, void *appdata_ptr)
{
    struct ad_user *user= &user_info;
#else
int PAM_conv (int num_msg, const struct pam_message **msg,
     struct pam_response **resp, void *appdata_ptr)
{
    struct ad_user *user= (struct ad_user *)appdata_ptr;
#endif /* PAM_SOLARIS_26 */
    struct pam_response *response;
    int i;

    /* Sanity checking */
    if (msg == NULL || resp == NULL || user == NULL)
    	return PAM_CONV_ERR;

#ifdef PAM_SOLARIS
    if (*msg == NULL)
	return PAM_CONV_ERR;
#endif

    response= (struct pam_response *)
    	malloc(num_msg * sizeof(struct pam_response));

    for (i= 0; i < num_msg; i++)
    {
	response[i].resp_retcode= 0;
	response[i].resp= NULL;

	switch (msgi(i).msg_style)
	{
	case PAM_PROMPT_ECHO_ON:
	    /* Store the login as the response */
	    /* This likely never gets called, since login was on pam_start() */
	    response[i].resp= appdata_ptr ? (char *)strdup(user->login) : NULL;
	    break;

	case PAM_PROMPT_ECHO_OFF:
	    /* Store the password as the response */
	    response[i].resp= appdata_ptr ? (char *)strdup(user->passwd) : NULL;
	    break;

	case PAM_TEXT_INFO:
	case PAM_ERROR_MSG:
	    /* Shouldn't happen since we have PAM_SILENT set. If it happens
	     * anyway, ignore it. */
	    break;

	default:
	    /* Something strange... */
	    if (response != NULL) free(response);
	    return PAM_CONV_ERR;
	}
    }
    /* On success, return the response structure */
    *resp= response;
    return PAM_SUCCESS;
}


/* CHECK_AUTH - Check a login and return a status code.
 * (This version for systems using PAM.)
 */

int check_auth(const char *login, const char *passwd)
{
#ifndef PAM_SOLARIS_26
    struct ad_user user_info= {login, passwd};
#endif /* PAM_SOLARIS_26 */
    struct pam_conv conv= { PAM_conv, (void *)&user_info };
    pam_handle_t *pamh= NULL;
    int retval;
    char *rhost;

#ifdef NEED_UID
    struct passwd *pwd;

    if ((pwd= getpwnam(login)) == NULL) return STATUS_UNKNOWN;
    hisuid= pwd->pw_uid;
    haveuid= 1;
#endif
#ifdef MIN_UNIX_UID
    if (hisuid < MIN_UNIX_UID) return STATUS_BLOCKED;
#endif

#ifdef PAM_SOLARIS_26
    user_info.login= login;
    user_info.passwd= passwd;
#endif /* PAM_SOLARIS_26 */

    /*
     * mod_authnz_external provides remote host information in optional
     * environment variables:
     *   IP: the IPv4 or IPv6 address
     *   HOST: the reverse resolution of the address
     * If present, we use one these to set the PAM_RHOST item,
     * preferring IP address over HOST.
     */
    rhost= getenv("IP");
    if (rhost == NULL)
	rhost= getenv("HOST");

    retval= pam_start("pwauth", login, &conv, &pamh);

    if (retval == PAM_SUCCESS && rhost != NULL)
	retval= pam_set_item(pamh, PAM_RHOST, rhost);

    if (retval == PAM_SUCCESS)
	retval= pam_authenticate(pamh, PAM_SILENT);

    if (retval == PAM_SUCCESS)
	retval= pam_acct_mgmt(pamh, 0);		/* permitted access? */

    if (pam_end(pamh,retval) != PAM_SUCCESS)	/* close PAM */
    {
	pamh= NULL;
	return STATUS_INT_ERR;
    }

    /* On failure we always return STATUS_UNKNOWN although we can't tell
     * if the failure was because of a bad login or a bad password */
    return (retval == PAM_SUCCESS ? STATUS_OK : STATUS_UNKNOWN);
}
#endif  /* PAM */


checkfaillog.c
/* =======================================================================
 * Copyright 1996, Jan D. Wolter and Steven R. Weiss, All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The names of the authors must not be used to endorse or
 *    promote products derived from this software without prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
 * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * =======================================================================
 */

#include <stdio.h>
#include <sys/types.h>
#include <utmp.h>
#include <pwd.h>
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include "config.h"
#include "fail_log.h"


int main(int argc, char **argv)
{
    int i, j;
    int reset= 0, verbose= 1;
    char *user= NULL;
    const char *msg= NULL;
    uid_t uid= getuid();

    /* Parse command line */
    for (i= 1; i < argc; i++)
    {
	if (argv[i][0] == '-')
	{
	    for (j= 1; argv[i][j] != '\0'; j++)
	    {
		switch (argv[i][j])
		{
		    case 'z': reset= 1; break;
		    case 's': verbose= 0; break;
		    default: goto usage;
		}
	    }
	}
	else if (user == NULL)
	    user= argv[i];
	else
	    goto usage;
    }

    /* Root can run this on other user's accounts */
    if (user != NULL)
    {
	struct passwd *pw;

	if ((pw= getpwnam(user)) == NULL)
	{
	    fprintf(stderr,"%s: User %s does not exist.\n", argv[0], user);
	    exit(2);
	}

	if (uid != 0 && pw->pw_uid != uid)
	{
	    fprintf(stderr,"%s: Only root can access other user's accounts.\n",
		    argv[0]);
	    exit(3);
	}

	uid= pw->pw_uid;
    }

    /* Do the thing */
#if defined(FAILLOG_JFH) || defined(FAILLOG_OPENBSD)
    msg= check_fails(uid, reset, verbose);
#endif

    /* Output the result */
    if (msg != NULL)
    	puts(msg);
    else if (!verbose)
    	puts("0:::");

    exit(0);

usage: fprintf(stderr,"usage: %s [-z] [-s] [user]\n", argv[0]);
    exit(1);
}


config.h
/* =======================================================================
 * Copyright 1996, Jan D. Wolter and Steven R. Weiss, All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The names of the authors must not be used to endorse or
 *    promote products derived from this software without prior written
 *              permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
 * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * =======================================================================
 */


/* PWAUTH configuration file 
 *
 * Note - the default settings in this file are the way I use it.  I
 * guarantee you, they won't work for you.  You must change them.
 *
 */


/* There are lots of different kinds of password databases.  Define one of
 * the following:
 *
 *  - SHADOW_NONE: old fashioned Unix systems which have the encrypted
 *    passwords in the passwd file.  Actually, since you don't need to be
 *    root to access these, you might do better using Lou Langholtz's
 *    mod_auth_system.c, which is available from the contributed modules
 *    directory at apache.org.
 *
 *  - SHADOW_BSD: This is the shadow password system in BSDI, NetBSD, OpenBSD,
 *    and FreeBSD.  It uses exactly the same calls as SHADOW_NONE, but
 *    getpwnam() only returns the encrypted password if you are root.  This
 *    would only work with OS X if the accounts being authenticated are
 *    configured with legacy crypt style passwords.  In general, the PAM
 *    option is more likely to be usuable in OS X.
 *
 *  - SHADOW_SUN:  This is the shadow password system in Solaris, Linux and
 *    IRIX 5.3 systems.  It uses getspnam() to fetch passwords and crypt() to
 *    encrypt them.
 *
 *  - SHADOW_JFH:  This the old version of Julianne F. Haugh's public-domain
 *    shadow package.  It uses getspnam() to fetch passwords and pw_encrpyt()
 *    to encrypt them.  The JFH shadow code is available at
 *        ftp://gumby.dsd.trw.com/pub/security/shadow-3.2.2.tar.Z
 *    Newer versions seem to be compatible with SHADOW_SUN.
 *
 *  - SHADOW_MDW:  This is Grex's variation on the JFH shadow file system,
 *    which uses Marcus D. Watt's interface to the password encryption.  If you
 *    ain't Grex, you ain't got it.
 *
 *  - SHADOW_AIX:  This is the AIX shadow password system.  It uses getuserpw()
 *    to fetch passwords and (apparantly) crypt() to encrypt them. This has
 *    not been tested.  Shadow BSD is also likely to work with AIX. The
 *    AUTHENTICATE_AIX option is probably a better option for AIX users.
 *
 *  - SHADOW_HPUX:  This is the HP-UX shadow password system.  It uses
 *    getprpwnam() to fetch passwords and either crypt() or bigcrypt() to
 *    encrypt them (I'm not sure which is right).  This has not been tested
 *    and probably doesn't work.
 *
 *  - PAM: Talk to the authentication system through PAM - the plug-in
 *    authentication module interface.  This exists on Solaris 7, FreeBSD,
 *    most versions of Linux and OS X.  You'll need to create
 *    /etc/pam.d/pwauth or edit /etc/pam.config to include entries for
 *    pwauth.  If you are using PAM to authenticate out of something you
 *    don't need to be root to access, then you might use instead Ingo
 *    Lutkebohle's mod_auth_pam.c module.  You probably want to comment
 *    out the UNIX_LASTLOG, MIN_UNIX_UID, and NOLOGIN_FILE options below.
 *
 *  - PAM_SOLARIS:  Solaris versions of PAM have some incompatibilities with
 *    Linux/FreeBSD versions.  I think HP-UX needs this too, but don't really
 *    know.
 *
 *  - PAM_SOLARIS_26:  Solaris 2.6 PAM has some header file declarations and
 *    function behaviors that don't agree with either their own documentation
 *    or with any other implementation.  Bugs, in short.  This option uses
 *    PAM with work-arounds for the Solaris 2.6 bugs.
 *
 *  - PAM_OS_X:  OS X keeps it's header files in a different place, so use
 *    this option instead of the PAM option.
 *
 *  - LOGIN_CONF_OPENBSD:  Many BSD derived systems use a login.conf file to
 *    configure authentication instead of (or in addition to) PAM.  We
 *    currently support authentication through this mechanism only for
 *    OpenBSD.  Of course, if your login.conf configuration is standard, you
 *    can just use SHADOW_BSD, but if you want pwauth to respect settings
 *    in login.conf this option can be used instead. The API used here, is
 *    however, pretty much unique to OpenBSD and will not work on NetBSD or
 *    FreeBSD.
 *
 *  - AUTHENTICATE_AIX: AIX has it's own system for configuring authentication
 *    via various files in the /etc/security directory. This can be used to
 *    configure special authentication parameters on a per-user basis including
 *    things like authenticating via kerberos and ldap and such like. We can
 *    tie into this interface via the authenticate() system call. The module
 *    to support this was contributed by a user and has not been tested by
 *    the author.
 */

   /* LOW-LEVEL OPTIONS */

/* #define SHADOW_NONE		/**/
/* #define SHADOW_BSD		/* FreeBSD, NetBSD, OpenBSD, BSDI, OS X */
/* #define SHADOW_SUN		/* Linux, Solaris, IRIX */
/* #define SHADOW_JFH		/**/
/* #define SHADOW_MDW		/**/
/* #define SHADOW_AIX		/* AIX (see also AUTHENTICATE_AIX) */
/* #define SHADOW_HPUX		/* HPUX ? */

   /* HIGH-LEVEL OPTIONS */

/* #define PAM			/* Linux PAM or OpenPAM */
/* #define PAM_OLD_OS_X		/* PAM on OS X version 10.5 or older */
/* #define PAM_SOLARIS		/* PAM on Solaris other than 2.6 */
/* #define PAM_SOLARIS_26	/* PAM on Solaris 2.6 */
/* #define LOGIN_CONF_OPENBSD	/* login.conf on OpenBSD */
/* #define AUTHENTICATE_AIX	/* AIX authenticate() function */


/* There is also limited support for three failure logging systems (the
 * database that informs you that "there have been 3426 unsuccessful attempts
 * to log into your account since your last login" and which may disable
 * accounts with too many failed logins).
 *
 * If a FAILLOG option is enabled, pwauth will increment the failure count
 * each time there is a failed attempt to login.  Depending on the
 * configuration, it may also deny logins to users who have had too many
 * bad login attempts.
 *
 * Very few Unix systems seem to have faillog files installed, so most
 * installations will not want this option.
 *
 * No faillog option should be used with PAM or AUTHENTICATE_AIX.  This kind
 * of logging is handled at a lower level within those systems.
 *
 *  - FAILLOG_JFH:  This is associated with the JFH shadow system.  Some Linux
 *    systems may have this, but most don't seem to.  Failures are logged in
 *    the /var/adm/faillog file, and if any user accumulates too many failed
 *    logins, future logins are denied.  The faillog.h header file is part of
 *    the JFH shadow file package.  If you define PATH_FAILLOG, then this
 *    will be used as the path of the faillog file instead of the one defined
 *    in faillog.h.
 *
 *  - FAILLOG_OPENBSD:  OpenBSD has a faillog, although it does not disable
 *    logins if any maximum is exceeded.  Failure counts are kept in
 *    /var/log/failedlogin.  There is no system header file that defines the
 *    format of this file, however.  Instead the definition for the file
 *    format is embedded in the "login" source code.  Bad things will happen
 *    if the definition there does not the match the definition in pwauth.
 *
 *    Though OpenBSD's login program does not refuse further logins if the
 *    number of failed logins has gotten too large, pwauth will do so if you
 *    define MAX_FAIL_COUNT.  If you want to use a file different than the
 *    default, you can define this in PATH_FAILLOG.
 *
 *  - FAILLOG_PWAUTH:  This is meant to be used by systems that lack any
 *    native faillog.  This will keep a faillog for pwauth only.  For the
 *    moment, this is identical to FAILLOG_OPENBSD, except that the default
 *    path of the log file is different.  In future releases this may
 *    diverge from FAILLOG_OPENBSD.  Defining MAX_FAIL_COUNT and PATH_FAILLOG
 *    has the same effect as above.
 *
 * RESET_FAIL_COUNT controls whether or not the failure count is reset to
 * zero when a successful login occurs.  The advantages of this are obvious.
 * The problem with it is that it obliterates all record of how many failures
 * there were.  Login utilities normally print out the failure count before
 * resetting it, so that users are notified if their account is under attack,
 * but pwauth cannot print messages that the user will see.  The optimum
 * strategy would probably be to have your CGI run a separate program, such
 * as the 'checkfaillog' program included in this distribution, that
 * reports and resets the failure count.
 */

/* #define FAILLOG_JFH				/**/
/* #define FAILLOG_OPENBSD			/**/
/* #define FAILLOG_PWAUTH			/**/

/* #define PATH_FAILLOG "/var/log/faillog"	/**/
/* #define MAX_FAIL_COUNT 40			/**/
/* #define RESET_FAIL_COUNT			/**/


/* If UNIX_LASTLOG is defined, the program will update the lastlog entry so
 * that there is a record of the user having logged in.  This is important on
 * systems where you expire unused accounts and some users may only log in
 * via the web.  If you have the lastlog.h header file, define HAVE_LASTLOG_H.
 *
 * If you are using PAM to authentication out of something other than the
 * unix user database, then you should disable this.  The lastlog file is
 * index by the user's uid number, and users from other databases don't have
 * uid numbers.
 */

/* #define UNIX_LASTLOG		/**/
/* #define HAVE_LASTLOG_H		/**/


/* If NOLOGIN_FILE is defined to the full path name of a file, then the
 * existance of that file is checked on every login attempt.  If it exists
 * all logins to accounts with uid's of MIN_NOLOGIN_UID or more will be
 * rejected.
 *
 * If you are using PAM, then you should disable this.  In that case, this
 * check is better done through PAM, and the maximum uid check won't work
 * right with PAM.
 */

/* #define NOLOGIN_FILE "/etc/nologin"	/**/
/* #define MIN_NOLOGIN_UID 1		/**/


/* Defining CHECK_LOGIN_EXPIRATION and CHECK_PASSWORD_EXPIRATION causes
 * pwauth to check if the given login has expired, or it's password has
 * expired.  Most modern versions of Unix support these options.
 *
 * These checks have not been implemented for the SHADOW_AIX and SHADOW_HPUX
 * options.  It's probably possible to do so for SHADOW_HPUX, but lacking a
 * system to test on, I haven't bothered.
 */

/* #define CHECK_LOGIN_EXPIRATION		/**/
/* #define CHECK_PASSWORD_EXPIRATION	/**/


/* It is generally sensible to restrict what users can run pwauth.  Though
 * there are other programs that users can use to test if password guesses
 * are correct, like "su" and "passwd", pwauth is much easier to interface
 * a password guessing program to, so why not be paranoid and restrict it
 * as much as possible?
 *
 * If you are using pwauth with mod_auth_external, you will want to restrict
 * it to be runnable from whatever uid your httpd runs as.  (For apache, this
 * is determined by the "User" directive in your httpd.conf file.  It may be
 * called something like "nobody" or "httpd" or "apache".  Usually the easiest
 * way to figure it out is just to do a "ps" and see what most apache processes
 * are running as.)
 *
 * There are two ways to configure this.  First, you can compile in the uid
 * numbers that are allowed to run this program, by listing them on the
 * SERVER_UID variable below.  At runtime, pwauth will check that the uid
 * of the user that invoked it is on this list.  So if you have just one
 * uid that should be able to run pwauth, you can say something like:
 *    #define SERVER_UIDS 72
 * If you have several, separate them by commas, like this:
 *    #define SERVER_UIDS 12,343,93
 * The root account can always run this program, so you don't have to
 * to list that.  If you do list it, it must be given last.
 *
 * The second option is to create a special group, called something like
 * "pwauth" for user id's that are allowed to run pwauth.  To do this, you
 * should compile pwauth with the SERVER_UIDS variable UNDEFINED (that is
 * the definition should be removed or commented out).  This will disable the
 * runtime uid check.  Then, when you install the pwauth program, set it's
 * group ownership to the "pwauth" group, and permit it so that only the
 * owner and the group can run it.  Do not permit it to be executable to
 * others.  This has the advantage of not requiring a recompile if you want
 * to change the uid list.
 */

/* #define SERVER_UIDS 30		/* user "wwwrun" on the author's system */


/* If MIN_UNIX_UID is defined to an integer, logins with uid numbers less than
 * that value will be rejected, even if the password is correct.
 *
 * If you are using PAM to authenticate against anything other than the local
 * unix password file, then leave this undefined (if you define it, only login
 * names which are in the local unix passwd file and have uid's greater the
 * given value will be accepted).
 */

/* #define MIN_UNIX_UID 500	/**/


/* If IGNORE_CASE is defined, the login given is checked in two different
 * ways. First without any changes and then with all letters converted to
 * lower case. This is useful for users accustomed to the Windows environment.
 * This ignores the case of the login name only, not the password.
 */

/* #define IGNORE_CASE             /**/


/* If DOMAIN_AWARE is enabled, then we we check login names to see if they
 * contain a backslash, and discard anything up to and including the backslash.
 * This is for use in environments where there are Windows users accustomed
 * to login names formed like "domain\username".
 */

/* #define DOMAIN_AWARE            /**/


/* On failed authentications, pwauth will sleep for SLEEP_TIME seconds, using
 * a lock on the file whose full path is given by SLEEP_LOCK to prevent any
 * other instances of pwauth from running during that time.  This is meant
 * to slow down password guessing programs, but could be a performance
 * problem on extremely heavily used systems.  To disable this, don't define
 * SLEEP_LOCK.  SLEEP_TIME defaults to 2 seconds if not defined.
 */

/* #define SLEEP_LOCK "/var/run/pwauth.lock"


/* If ENV_METHOD is defined, pwauth expects mod_auth_external to be configured
 *     SetAuthExternalMethod environment
 * instead of
 *     SetAuthExternalMethod pipe
 * This is insecure on some versions of Unixes, but might be a bit faster.
 */

/* #define ENV_METHOD		/**/


/* If /usr/include/paths.h exists define this.  Obviously I need to autoconfig
 * this.
 */

/* #define PATHS_H 		/**/
